### PR TITLE
Specify Oak Application format

### PR DESCRIPTION
**Pull request by tiziano88**
Implement only support for applications with a single node and no
channels for now; in the future we should remove the current hardcoded
channels and specify them in the application configuration instead.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -87,7 +87,7 @@ machine.
 The unit of compilation and execution in Oak is an **Oak Module**. Each Oak
 Module is a self-contained
 [WebAssembly module](https://webassembly.org/docs/modules/) that is interpreted
-by an Oak VM instance as part of an Oak Node.
+by an Oak VM instance as part of an Oak Application.
```
Comment (anghelcovici) [inline README.md]: I am wondering if calling it a Oak VM might lead people to expect certain things. Is it actually a VM?

File: examples/machine_learning/client/BUILD [hunk 2]
```diff
@@ -20,8 +20,8 @@ cc_binary(
     deps = [
         "//examples/machine_learning/proto:machine_learning_cc_grpc",
         "//examples/utils",
+        "//oak/client:application_client",
         "//oak/client:manager_client",
```
Comment (anghelcovici) [inline examples/machine_learning/client/BUILD]: Manager_client sounds strange :) It makes sense though

File: README.md [hunk 3]
```diff
@@ -179,22 +179,7 @@ Rust abstractions over the lower level WebAssembly interface.
...
-running a genuine Oak VM and will therefore obey the policies set on the Oak
-Node; the Oak VM itself may then optionally prove additional details about the
-Oak Module and its properties, which may require reasoning about its internal
-structure.
-
-The Oak Module and the policies associated with an Oak Node are established once
-and for all at the time the Oak Node is created by the Oak Manager, and they
-cannot be modified once the Oak Node is running. Therefore each client only
-needs to verify the attestation once before it starts invoking the Oak Node.
+An **Oak Node** is an instance of an Oak Module running on an Oak VM.
```
Comment (anghelcovici) [inline README.md]: So we have 1:1 Module:Node and n:1 Nodes/Modules:Application? Do we really need the distinction between a node and a module? (similar to object/instance?)

File: oak/client/manager_client.h [hunk 4]
```diff
@@ -28,17 +28,21 @@ class ManagerClient {
...
+  // TODO: Return StatusOr<::oak::CreateApplicationResponse>.
+  CreateApplicationResponse CreateApplication(const std::string& module_bytes) {
     grpc::ClientContext context;
 
-    CreateNodeRequest request;
-    CreateNodeResponse response;
+    CreateApplicationRequest request;
+    CreateApplicationResponse response;
 
-    request.set_module(module_bytes);
```
Comment (daviddrysdale) [inline oak/client/manager_client.h]: Maybe add a comment here like
```c++
// Build an application configuration with a single Node and no channels.
// TODO: replace with gRPC and Log pseudo-Nodes (& associated channels) when available.
```

File: README.md [hunk 5]
```diff
@@ -211,22 +196,38 @@ invocations, but only if it can also be shown that the data can only be
...
 clients.
 
+## Oak Application
+
+An **Oak Application** is a set of Oak Nodes running within the same enclave,
+and connected by unidirectional channels. The connectivity graph is specified as
+part of an [Application Configuration](/oak/proto/manager.proto) and is
+immutable once an application is running.
+
+An Oak Application may have one of more entry points from which it can be
```
Comment (daviddrysdale) [inline README.md]: s/of/or/

File: README.md (hunk 1 above)
Comment (tiziano88) [reply README.md]: Yeah I think we should just call it Oak Node or something like that and drop the term VM entirely.

File: README.md (hunk 3 above)
Comment (tiziano88) [reply README.md]: I am using Module to refer to the code, and Node to the instance. In this terminology, developers write Modules, not Nodes. But if that's confusing we can drop the term Module too and just say that developers write Nodes, even if it's not super accurate. Thoughts?

File: README.md (hunk 5 above)
Comment (tiziano88) [reply README.md]: Done.

File: examples/machine_learning/client/BUILD (hunk 2 above)
Comment (tiziano88) [reply examples/machine_learning/client/BUILD]: Maybe deployment_client or similar?

File: oak/client/manager_client.h (hunk 4 above)
Comment (tiziano88) [reply oak/client/manager_client.h]: Good point, thanks.

File: oak/proto/manager.proto [hunk 6]
```diff
@@ -18,26 +18,86 @@ syntax = "proto3";
 
 package oak;
 
-message CreateNodeRequest {
+// An ApplicationConfiguration represents a unit of deployment in Oak.
+//
+// It defines a connectivity graph using unidirectional channels between nodes.
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/proto/manager.proto]: Done (also for Channels)

File: oak/proto/manager.proto [hunk 7]
```diff
@@ -18,26 +18,86 @@ syntax = "proto3";
...
+  }
+}
+
+// A node that corresponds to an external-facing gRPC server interface exposed
+// by the application. External clients can connect to the server and interact
+// with the application over gRPC.
+message GrpcServerNode {
+}
+
+// A node that logs anything that is sent to it to stdout on the machine on
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/proto/manager.proto]: Done.

File: oak/proto/manager.proto [hunk 8]
```diff
@@ -18,26 +18,86 @@ syntax = "proto3";
...
+}
+
+message WebAssemblyNode {
   // The compiled code of the Oak Module to instantiate, in WebAssembly binary
   // format.
   // See https://webassembly.org/docs/binary-encoding/ .
-  bytes module = 1;
+  // TODO: Replace this with just a hash of the bytecode to instantiate, and
+  // pass the actual bytecode to the Oak Manager in some other way.
+  bytes module_bytes = 3;
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/proto/manager.proto]: Magic!

File: oak/proto/manager.proto [hunk 9]
```diff
@@ -18,26 +18,86 @@ syntax = "proto3";
...
+  string node_name = 1;
+
+  oneof node_type {
+    LogNode log_node = 2;
+    GrpcServerNode grpc_server_node = 3;
+    WebAssemblyNode web_assembly_node = 4;
+  }
+}
+
+// A node that corresponds to an external-facing gRPC server interface exposed
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/proto/manager.proto]: Done, though I added the constraint at the Application level.

File: oak/proto/manager.proto [hunk 10]
```diff
@@ -18,26 +18,86 @@ syntax = "proto3";
...
+// by the application. External clients can connect to the server and interact
+// with the application over gRPC.
+message GrpcServerNode {
+}
+
+// A node that logs anything that is sent to it to stdout on the machine on
+// which the application is running.
+// This is unsafe and only to be used for testing, since it leaks all the data
+// that it handles.
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/proto/manager.proto]: Added constraint of at most 1 LogNode in the Application comment.

**Pull request merged**